### PR TITLE
virsh_freepages: add new cases and more strictly result checking

### PR DESCRIPTION
**Pull request by czhang0**
virsh_freepages: add new cases and more stictly result checking
    
    New cases:
    1. cover 3 type pagesizes testing, ranging from 4K, 2M, 1G
    2. cover severl pagesize formates
       4 is same with 4K, 4k, 4Kib, 4KIB
       2048 is same with 2048K, 2M, 2m, 2048Kib and so on
       1048576 is same with 1G, 1048576k, 1048576Kib and so on
    
    New checking points:
    1. Check if hugepages feature is supported by CPU
    2. Check if hugepages can be available by current system
    3. Check allocated pagesize number for each available pagesize
    4. Compare free pagesize number with allocated pagesize number for
       each available  pagesize
    
    Signed-off-by: cuzhang <cuzhang@redhat.com>

### CODE REVIEW

**Comment by czhang0** [comment]
Run case result on local:
#avocado run --vt-type libvirt type_specific.local.virsh.freepages
JOB ID     : 6c59d62cc16abace19826b87d7648c57ec9cd61c
JOB LOG    : /root/avocado/job-results/job-2018-01-08T11.31-6c59d62/job.log
 (01/25) type_specific.local.virsh.freepages.positive_test.no_pagesize.option_all: PASS (3.21 s)
 (02/25) type_specific.local.virsh.freepages.positive_test.no_pagesize.each_cells: PASS (4.22 s)
 (03/25) type_specific.local.virsh.freepages.positive_test.each_pagesize.option_all: PASS (3.77 s)
 (04/25) type_specific.local.virsh.freepages.positive_test.each_pagesize.each_cells: PASS (4.25 s)
 (05/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_1.option_all: PASS (3.69 s)
 (06/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_1.each_cells: PASS (4.57 s)
 (07/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_2.option_all: PASS (3.62 s)
 (08/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_2.each_cells: PASS (4.53 s)
 (09/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_3.option_all: PASS (3.59 s)
 (10/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_3.each_cells: PASS (4.54 s)
 (11/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_4.option_all: PASS (3.58 s)
 (12/25) type_specific.local.virsh.freepages.positive_test.2M_pagesize_4.each_cells: PASS (4.54 s)
 (13/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_1.option_all: PASS (3.58 s)
 (14/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_1.each_cells: PASS (4.70 s)
 (15/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_2.option_all: PASS (3.59 s)
 (16/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_2.each_cells: PASS (4.64 s)
 (17/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_3.option_all: PASS (3.59 s)
 (18/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_3.each_cells: PASS (4.53 s)
 (19/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_4.option_all: PASS (4.53 s)
 (20/25) type_specific.local.virsh.freepages.positive_test.1G_pagesize_4.each_cells: PASS (4.51 s)
 (21/25) type_specific.local.virsh.freepages.negative_test.invalid_option: PASS (2.73 s)
 (22/25) type_specific.local.virsh.freepages.negative_test.mutually_exclusive_option: PASS (2.51 s)
 (23/25) type_specific.local.virsh.freepages.negative_test.out_of_range_cellno: PASS (1.60 s)
 (24/25) type_specific.local.virsh.freepages.negative_test.invalid_cellno: PASS (2.58 s)
 (25/25) type_specific.local.virsh.freepages.negative_test.invalid_page_size: PASS (1.58 s)
RESULTS    : PASS 25 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 107.10 s

File: libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg [hunk 1]
```diff
@@ -7,18 +7,30 @@
...
+                - each_cells:
+                    freepages_cellno = "EACH"
             variants:
-                - no_hugepage:
-                - default_hugepage:
-                    huge_pages_num = "1"
+                - no_pagesize:
+                - each_pagesize:
+                    freepages_pagesize = "EACH"
+                - 2M_pagesize_1:
```
Comment (chunfuwen) [inline libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg]: rename to pagesize_in_M

File: libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg [hunk 2]
```diff
@@ -7,18 +7,30 @@
...
+                    freepages_pagesize = "2M"
+                - 2M_pagesize_2:
```
Comment (chunfuwen) [inline libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg]: rename to pagesize_in_default, other variables have similar issues, better outstand unit in variable name

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 3]
```diff
@@ -48,76 +49,148 @@ def check_freepages(output, expect_result_list):
...
-            if expect == all_freepages[0]:
+            cell = all_freepages[0]
+            modify_expect(cell, expect)
+            if expect == cell:
                 check_pass = True
         check_result.append(check_pass)
     return False not in check_result
 
 
+def modify_expect(cell, expect, pagesize='4KiB'):
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: miss parameter doc string

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 4]
```diff
@@ -48,76 +49,148 @@ def check_freepages(output, expect_result_list):
...
+
+
 def run(test, params, env):
     """
     Test virsh command virsh freepages
     """
 
+    def get_node_freepages(node, pagesize):
+        """
+        Get number of hugepages for node
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: add one blank line here

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 5]
```diff
@@ -48,76 +49,148 @@ def check_freepages(output, expect_result_list):
...
+        :params node: node No
+        :params pagesize: page size, for example: '4', '2048', '1048576'
+            For pagesize is '4' which is not hugepage, get free memory from
+                `grep MemFree /sys/devices/system/node/node${node}/meminfo`
+            For pagesize '2048' or '1048576',  get free page number from
+                /sys/devices/system/node/nodei${node}/hugepages/
+                hugepages-${pagesize}/free_hugepages
+        :params return: integer of free pagesize
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: return need one single line

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 6]
```diff
@@ -48,76 +49,148 @@ def check_freepages(output, expect_result_list):
...
+        """
+        if pagesize == '4':
+            node_meminfo = host_numa_node.read_from_node_meminfo(node, 'MemFree')
+            return int(node_meminfo) / 4
+        else:
+            return hp_cl.get_node_num_free_pages(node, pagesize)
+
+    def check_supported_status(check_type, pagesize, quantity):
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: miss parameter doc string

File: libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg (hunk 1 above)
Comment (czhang0) [reply libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg]: Modified according to your advice:
 18                 - pagesize_2M:
 19                     freepages_pagesize = "2M"
 20                 - pagesize_2048:
 21                     freepages_pagesize = "2048"
 22                 - pagesize_2048K:
 23                     freepages_pagesize = "2048K"
 24                 - pagesize_2048kib:
 25                     freepages_pagesize = "2048kib"
 26                 - pagesize_1G:
 27                     freepages_pagesize = "1G"
 28                 - pagesize_1048576:
 29                     freepages_pagesize = "1048576"
 30                 - pagesize_1048576K:
 31                     freepages_pagesize = "1048576K"
 32                 - pagesize_1048576KIB:
 33                     freepages_pagesize = "1048576KIB"

File: libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg (hunk 2 above)
Comment (czhang0) [reply libvirt/tests/cfg/virsh_cmd/host/virsh_freepages.cfg]: Modify to use more meaningful name. Thanks.

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py (hunk 4 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: Adopt it. Thanks.

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py (hunk 5 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: Adopt it. Thanks.

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py (hunk 6 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: Added. Thanks!

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 7]
```diff
@@ -48,76 +50,156 @@ def check_freepages(output, expect_result_list):
...
+        if abs(int(expect[pagesize]) - int(cell[pagesize])) < 1024:
+            expect[pagesize] = cell[pagesize]
+
+
 def run(test, params, env):
     """
     Test virsh command virsh freepages
     """
 
+    def get_node_freepages(node, pagesize):
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: rename to get_freepages_number_of_node, maybe more meaningful

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 8]
```diff
@@ -48,76 +50,156 @@ def check_freepages(output, expect_result_list):
...
+        """
+        if pagesize == '4':
+            node_meminfo = host_numa_node.read_from_node_meminfo(node, 'MemFree')
+            return int(node_meminfo) / 4
+        else:
+            return hp_cl.get_node_num_free_pages(node, pagesize)
+
+    def check_supported_status(check_type, pagesize, quantity):
+        """
+        Allocate some hugepages and get the actuall number, return status
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: maybe update it as : 
Allocate some hugepages to specfic sizes ,then check whether those huge pages size is equals to free pages size of the sum of all nodes

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 9]
```diff
@@ -48,76 +50,156 @@ def check_freepages(output, expect_result_list):
...
+                    logging.debug('move node %s out of testing node list', node)
+                    node_list.remove(node)
+                    oth_node.append(node)
+                if allocate_pages != free_pages:
+                    # Check if allocated hugepages is equal to freepages for
+                    # each node
+                    test.error("allocated hugepages is not equal to "
+                               "free pages.")
+            if hp_sum != actual_quantity:
+                test.error("the sum of 1G hugepages for each nodes is not "
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: maybe not hardcard string '1G hugepages'

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py [hunk 10]
```diff
@@ -48,76 +50,156 @@ def check_freepages(output, expect_result_list):
...
+                hugepages-${pagesize}/free_hugepages
+        :return: integer number of free pagesize
+        """
+        if pagesize == '4':
+            node_meminfo = host_numa_node.read_from_node_meminfo(node, 'MemFree')
+            return int(node_meminfo) / 4
+        else:
+            return hp_cl.get_node_num_free_pages(node, pagesize)
+
+    def check_supported_status(check_type, pagesize, quantity):
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: rename method name to check_hugepages_allocated_status,and rename check_type to is_type_check to make more meaningful.

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py (hunk 7 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: Yes, get_freepages_number_of_node is better. Got it.

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py (hunk 8 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: Modified according to your advice:
Allocate some specific hugepages and check whether hugepages number
allocated by kernel totally is equal to the sum of allocated for each
nodes, and also check if the number of freepages for this specific 
hugepages for each node is equal to the number allocated by 
kernel for each node. If yes, return True, else return False.

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py (hunk 9 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: Modified to use variable pagesize here. Thanks.

File: libvirt/tests/src/virsh_cmd/host/virsh_freepages.py (hunk 10 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/host/virsh_freepages.py]: Yes, your name is better to understand, adopt it, thanks!

**Pull request merged**